docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,240 @@
+
+    	// Env are environment variables, in the `key=value` form, passed to the build.
+    	// For example: `CGO_ENABLED=1`.
+    	Env []string `yaml:"env,omitempty"`
+
+    	// Platforms is the list of platforms to build images for. Each platform
+    	// is of the format `os/arch[/variant]`, e.g., `linux/amd64`.
+    	// By default, the ko builder builds for `all` platforms supported by the
+    	// base image.
+    	Platforms []string `yaml:"platforms,omitempty"`
+
+    	// SourceDateEpoch is the `created` time of the container image.
+    	// Specify as the number of seconds since January 1st 1970, 00:00 UTC.
+    	// You can override this value by setting the `SOURCE_DATE_EPOCH`
+    	// environment variable.
+    	SourceDateEpoch unit64 `yaml:"sourceDateEpoch,omitempty"`
+    }
+    ```
+
+3.  Add a `KoArtifact` field to the `ArtifactType` struct:
+
+    ```go
+    type ArtifactType struct {
+    	// KoArtifact builds images using [ko](https://github.com/google/ko).
+    	KoArtifact *KoArtifact `yaml:"ko,omitempty" yamltags:"oneOf=artifact"`
+    }
+    ```
+
+Example basic config, this will be sufficient for many users:
+
+```yaml
+apiVersion: skaffold/v2beta14
+kind: Config
+build:
+  artifacts:
+  - image: ko://example.com/helloworld
+    ko: {}
+```
+
+The value of the `image` field is the Go import path of the app entry point,
+[prefixed by `ko://`](https://github.com/google/ko/pull/58).
+
+ko requires setting a
+[`KO_DOCKER_REPO`](https://github.com/google/ko#choose-destination)
+environment variable. This determines where container images are pushed.
+The Skaffold
+[default repo](https://skaffold.dev/docs/environment/image-registries/)
+maps directly to this value.
+
+### Open Questions
+
+1.  Should Skaffold embed ko (as a Go module), or shell out?
+
+    Benefits of embedding:
+
+    - Skaffold can pin the ko version it supports in its `go.mod` file. Users
+      wouldn't raise bugs/issues for incompatible version pairings of Skaffold
+      and ko.
+
+    - Reduce toolchain maintenance toil for users. Skaffold users wouldn't need
+      to synchronize ko versions used by different team members or in their CI
+      build, since the Skaffold version determines the ko version.
+
+    - Portability. Skaffold+ko users only need one tool for their container
+      image building needs: the `skaffold` binary. (Plus the Go distribution,
+      of course.) The current `gcr.io/k8s-skaffold/skaffold` container image
+      could serve as a build and deploy image for CI/CD pipeline steps.
+
+    Embedding ko would require some level of documented behavioural stability
+    guarantees for the most ko interfaces that Skaffold would use, such as
+    [`build.Interface`](https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/build/build.go#L24)
+    and
+    [`publish.Interface`](https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/publish/publish.go#L24),
+    or others?
```

Comment:
Wouldn't there be some code duplication across skaffold and ko repos to create the publisher and builder object. All of this https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/commands/publish.go#L62

Reply:
Agree, there would be. These seemed like the most appropriate existing interfaces. We should look into defining something at a slightly higher level in ko to avoid the duplication.

Reply:
Would you take up that work too ?

Reply:
Yes, I'll coordinate with the ko team and create the necessary PRs to do that.